### PR TITLE
support of multiple devices for tvm.build

### PR DESCRIPTION
**Pull request by zhiics**
This PR mainly focuses on  supporting multiple devices for `tvm.build`, it is part of #1688.

The proposal 2 of #1752 is implemented to support multiple devices for compilation and code generation. In the end, only one module will be generated.  (In compiler PR, in order to pass `target: loweredfunc_list` pairs to `tvm.build`, we need to slightly modify the `GraphComplie` pass. Multiple targets need to be glued to the graph for compiling, and a `device_type` field may need to be added   to`NodeAttr` or `Op`. )

### CODE REVIEW

**Comment by srkreddy1238** [comment]
@zhiics 

Ref. RFC #1242 
- It talks about ```FComputeFallback``` which indicates the operator to have a fallback to CPU.

Here the device_types is selected based on the result of ```FComputeFallBack``` not by explicit arguments to nnvm build. We could continue to have targets passed to nnvm build to control the fall back devices.
Am I missing anything here ?

How about another PR first to demonstrate the fallback behavior of operator and copy insertion before integrating nnvm build to tvm build.

**Comment by zhiics** [comment]
@srkreddy1238 

device_types is not selected by explicit argument to nnvm build here. It is determined through annotation which will be another PR. I think `FComputeFallBack` is registered to tell the compiler that the operator should fallback to the host (cpu). This could be done at the annotation pass. I did this first is because I think annotation may need more discussion.

Or I can remove the compiler related stuff now, but only keep the new `tvm.build` part to reduce confusion for this PR. How do you think?

File: python/tvm/build_module.py [hunk 1]
```diff
@@ -379,24 +379,40 @@ def lower(sch,
...
     """Build a function with arguments as signature. Code will be generated
-    for a device specified by the target. For homogeneous execution, a module
-    that contains both host and device code is returned. For heterogeneous
-    execution, a list of lowered functions for the host and a module containing
-    device code are returned, but actual code generation for the host module is
-    postponed after code generation is finished for all devices.
+    for devices coupled with target information.
+
+    There are two typical uses of this functioin:
+        1. For backward compatibility
```
Comment (tqchen) [inline python/tvm/build_module.py]: This is a valid case of usage so we don't want to see for backward compatiblity.

Consider put these code examples under Example section with rst real code block

File: python/tvm/build_module.py [hunk 2]
```diff
@@ -379,24 +379,40 @@ def lower(sch,
         return stmt
     return ir_pass.MakeAPI(stmt, name, arg_list, 0, cfg.restricted_func)
 
+
 def build(sch,
```
Comment (tqchen) [inline python/tvm/build_module.py]: Let us rename sch to inputs, since now it is more than a scheduld

File: python/tvm/build_module.py [hunk 3]
```diff
@@ -451,76 +459,96 @@ def build(sch,
...
+        for x in flist:
+            if not isinstance(x, container.LoweredFunc):
+                raise ValueError("sch have to be Schedule, LoweredFunc, list "
+                                 "of LoweredFunc, or dict of str to list of "
+                                 "LoweredFunc.")
+            if x.name in fname_set:
+                raise ValueError("Duplicate function name %s" % x.name)
+            fname_set.add(x.name)
+
+    fhost_all = []
```
Comment (tqchen) [inline python/tvm/build_module.py]: Let us move the older per device logic to a separate internal function

**Comment by tqchen** [comment]
Thanks @srkreddy1238 @zhiics , this is now merged

**Pull request merged**